From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shelley Vohr <shelley.vohr@gmail.com>
Date: Mon, 27 Nov 2023 17:23:29 +0100
Subject: build: only create cppgc heap on non-32 bit platforms

Node.js crashes during heap marking when using cppgc on 32 bit platforms. Disable for now.

Crash occurs in the following stack:
* MarkingVisitorBase<ConcreteVisitor>::VisitEmbedderTracingSubClassWithEmbedderTracing
* MarkingWorklists::Local::PushExtractedWrapper
* CppMarkingState::MarkAndPush
* CppMarkingState::MarkAndPush
* WrappableInfo::From
* Crash

See https://gist.github.com/codebytere/275ec8923253fd6559b3d36115f7b31b for more.

This should be fixed and re-enabled on all platforms.

diff --git a/src/env.cc b/src/env.cc
index a429d5526d0af66eef94132b196e42994e3f6448..ba575a04340b91709fb6c8710ab160a4ca1f8b77 100644
--- a/src/env.cc
+++ b/src/env.cc
@@ -546,7 +546,8 @@ IsolateData::IsolateData(Isolate* isolate,
     // for embedder ID, V8 could accidentally enable cppgc on them. So
     // safe guard against this.
     DCHECK_NE(descriptor.wrappable_type_index, BaseObject::kSlot);
-  } else {
+  } else  {
+#if !defined(NODE_RUNNING_32_BIT)
     cpp_heap_ = CppHeap::Create(
         platform,
         CppHeapCreateParams{
@@ -554,6 +555,7 @@ IsolateData::IsolateData(Isolate* isolate,
             WrapperDescriptor(
                 BaseObject::kEmbedderType, BaseObject::kSlot, cppgc_id)});
     isolate->AttachCppHeap(cpp_heap_.get());
+#endif
   }
   // We do not care about overflow since we just want this to be different
   // from the cppgc id.
diff --git a/src/env.cc b/src/env.cc
index 1d8df40c3446ac9c72c5f0ad24edf0b8a96cf16b..25b81dee18aeeb1bd0452ba0b66085d451723709 100644
--- a/src/env.cc
+++ b/src/env.cc
@@ -542,7 +542,8 @@ IsolateData::IsolateData(Isolate* isolate,
     // for embedder ID, V8 could accidentally enable cppgc on them. So
     // safe guard against this.
     DCHECK_NE(descriptor.wrappable_type_index, BaseObject::kSlot);
-  } else {
+  } else  {
+#if !defined(NODE_RUNNING_32_BIT)
     cpp_heap_ = CppHeap::Create(
         platform,
         CppHeapCreateParams{
@@ -550,6 +551,7 @@ IsolateData::IsolateData(Isolate* isolate,
             WrapperDescriptor(
                 BaseObject::kEmbedderType, BaseObject::kSlot, cppgc_id)});
     isolate->AttachCppHeap(cpp_heap_.get());
+#endif
   }
   // We do not care about overflow since we just want this to be different
   // from the cppgc id.
=======
     if (is_win) {
       defines += [
         "NOMINMAX",
>>>>>>> 95a9029428 (chore: fix patch indices)
